### PR TITLE
[FE-20] FEAT: Implement contact us logic

### PR DESCRIPTION
**Pull request by alexkocharyan01**

### CODE REVIEW

**Review by ArthurO2002 — CHANGES_REQUESTED** [review]

File: src/redux/slices/support/initialState.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { ISupportProps } from '@allTypes/reduxTypes/supportStateTypes'
+
+export const getInitialState = (): ISupportProps => ({
+  support: {
+    error: null,
+    isHelpLoading: false,
+    isContactUsLoading: false,
+    contactUsStatus: null,
+  },
```
[the remark this answers is not in the document]
Comment (alexkocharyan01) [reply src/redux/slices/support/initialState.ts:9]: done

File: src/redux/slices/users/middleware.ts [hunk 2]
```diff
@@ -103,7 +103,7 @@ const changeLanguage = (lng: string) => async (dispatch: AppDispatch) => {
     dispatch(setLanguageChangeLoading(true))
     dispatch(setLanguage(lng))
 
-    i18n.changeLanguage(lng)
+    await i18n.changeLanguage(lng)
     localStorage.setItem('language', lng)
```
[the remark this answers is not in the document]
Comment (alexkocharyan01) [reply src/redux/slices/users/middleware.ts]: done

File: src/redux/slices/support/middleware.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    dispatch(setError(null))
+  } catch (error) {
+    dispatch(setError((error as IError).response.data.status.message))
+  } finally {
+    dispatch(setHelpLoading(false))
+  }
+}
+
+const fetchContactUsMessage = (message: IContactUsProps) => async (dispatch: AppDispatch) => {
+  try {
```
[the remark this answers is not in the document]
Comment (alexkocharyan01) [reply src/redux/slices/support/middleware.ts]: done

File: src/components/ContactUsSendMessage/index.tsx [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        phone: data.phone.toString(),
+        message: data.message,
+      })
+    )
+  }
+
+  useEffect(() => {
+    if (contactUsStatus === 200) {
+      reset({ ...defaultValues })
+    }
```
[the remark this answers is not in the document]
Comment (alexkocharyan01) [reply src/components/ContactUsSendMessage/index.tsx]: done

File: src/redux/slices/support/middleware.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import API from '@axios/API'
+import { IError } from '@axios/authentication/authManagerTypes'
+import { AppDispatch } from '@redux/store'
+
+import slice from './slice'
+
+const { setError, setHelpLoading, setContactUsLoading, setContactUsStatus } = slice.actions
+
+const fetchHelpMessage = (message: IHelpDataProps) => async (dispatch: AppDispatch) => {
+  try {
```
[the remark this answers is not in the document]
Comment (alexkocharyan01) [reply src/redux/slices/support/middleware.ts]: done

File: src/components/ContactUsSendMessage/index.tsx [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  }, [contactUsStatus, reset])
+
+  return (
+    <div className="xl:pr-20">
+      <form onSubmit={handleSubmit(onSubmit)}>
+        <div className="mb-5 flex flex-col">
+          <Controller
+            control={control}
+            name="fullName"
```
[the remark this answers is not in the document]
Comment (alexkocharyan01) [reply src/components/ContactUsSendMessage/index.tsx]: done

File: src/pages/contact-us/index.tsx [hunk 7]
```diff
@@ -1,95 +1,71 @@
...
-                id="telephone"
-                label="Telephone"
-                required
-              />
-            </div>
+const ContactUs = () => {
+  const [t] = useTranslation()
+
+  return (
+    <Container className="pb-30">
```
[the remark this answers is not in the document]
Comment (alexkocharyan01) [reply src/pages/contact-us/index.tsx:14]: done

File: src/redux/slices/support/middleware.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+const fetchContactUsMessage = (message: IContactUsProps) => async (dispatch: AppDispatch) => {
+  try {
+    dispatch(setContactUsLoading(true))
+
+    const response = await API.support.sendContactUsData(message)
+
+    dispatch(setContactUsStatus(response.status))
+
```
[the remark this answers is not in the document]
Comment (alexkocharyan01) [reply src/redux/slices/support/middleware.ts]: done

File: src/validation/support/contactUs.ts [hunk 9]
```diff
@@ -0,0 +1,8 @@
+import { object, string } from 'yup'
+
+export const contactUsValidationSchema = object({
+  fullName: string().required('Full name is required'),
+  email: string().email().required('Email is required'),
```
Comment (gor017) [inline src/validation/support/contactUs.ts:5]: this will not support correct validation of email, let's right manually regexp for it

**Review by gor017 — CHANGES_REQUESTED** [review]

File: src/validation/support/contactUs.ts [hunk 10]
```diff
@@ -0,0 +1,8 @@
...
+  phone: string().required('Phone Number is required'),
```
Comment (gor017) [inline src/validation/support/contactUs.ts:6]: phone should not include e

File: src/redux/slices/support/selector.ts [hunk 11]
```diff
@@ -3,8 +3,21 @@ import { createSelector } from '@reduxjs/toolkit'
 
 const selector = (state: RootState) => state.support
 
-export const support = createSelector([selector], (state) => state.support)
+export const help = createSelector([selector], (state) => state.support.help)
+export const isHelpMessageSuccess = createSelector(
+  [selector],
+  (state) => state.support.help.isHelpMessageSuccess
+)
```
Comment (gor017) [inline src/redux/slices/support/selector.ts:10]: change to isHelpDataSubmittedSuccess

File: src/redux/slices/support/selector.ts [hunk 12]
```diff
@@ -3,8 +3,21 @@ import { createSelector } from '@reduxjs/toolkit'
...
+export const contactUs = createSelector([selector], (state) => state.support.contactUs)
+
+export const isContactUsMessageSuccess = createSelector(
+  [selector],
+  (state) => state.support.contactUs.isContactUsMessageSuccess
+)
```
Comment (gor017) [inline src/redux/slices/support/selector.ts:16]: change to isContactUsDataSubmittedSuccess

File: src/types/reduxTypes/supportStateTypes.ts [hunk 13]
```diff
@@ -3,13 +3,30 @@ export interface ISupportProps {
...
   error: null | Error | string
   isHelpLoading: boolean
-
   isHelpMessageSuccess: boolean
 }
 
+export interface IContactUsPropsList {
+  error: null | Error | string
+  isContactUsLoading: boolean
+  isContactUsMessageSuccess: boolean
```
Comment (gor017) [inline src/types/reduxTypes/supportStateTypes.ts:19]: isContactUsDataSubmittedSuccess

File: src/types/reduxTypes/supportStateTypes.ts [hunk 14]
```diff
@@ -3,13 +3,30 @@ export interface ISupportProps {
...
+export interface IHelpPropsList {
   error: null | Error | string
   isHelpLoading: boolean
-
   isHelpMessageSuccess: boolean
 }
 
+export interface IContactUsPropsList {
+  error: null | Error | string
+  isContactUsLoading: boolean
```
Comment (gor017) [inline src/types/reduxTypes/supportStateTypes.ts:18]: isContactUsSubmittingLoading

File: src/types/reduxTypes/supportStateTypes.ts [hunk 15]
```diff
@@ -3,13 +3,30 @@ export interface ISupportProps {
 }
 
 export interface ISupportPropsList {
+  help: IHelpPropsList
+  contactUs: IContactUsPropsList
+}
+
+export interface IHelpPropsList {
   error: null | Error | string
   isHelpLoading: boolean
```
Comment (gor017) [inline src/types/reduxTypes/supportStateTypes.ts:12]: isHelpDataSublittingLoading

File: src/redux/slices/support/middleware.ts [hunk 16]
```diff
@@ -20,10 +26,36 @@ const createHelpMessage = (message: IHelpDataProps) => async (dispatch: AppDispa
...
+const resetContactUsMessageSuccess = () => (dispatch: AppDispatch) => {
+  dispatch(setContactUsMessageSuccess(false))
+}
+
 export default {
   createHelpMessage,
+  createContactUsMessage,
+  resetHelpMessageSuccess,
+  resetContactUsMessageSuccess,
```
Comment (gor017) [inline src/redux/slices/support/middleware.ts:60]: resetContactUsSubmitDataSuccess

File: src/redux/slices/support/middleware.ts [hunk 17]
```diff
@@ -20,10 +26,36 @@ const createHelpMessage = (message: IHelpDataProps) => async (dispatch: AppDispa
...
+}
+
+const resetContactUsMessageSuccess = () => (dispatch: AppDispatch) => {
+  dispatch(setContactUsMessageSuccess(false))
+}
+
 export default {
   createHelpMessage,
+  createContactUsMessage,
+  resetHelpMessageSuccess,
```
Comment (gor017) [inline src/redux/slices/support/middleware.ts:59]: resetHelpSubmitDataSuccess

File: src/redux/slices/support/reducers.ts [hunk 18]
```diff
@@ -6,13 +6,19 @@ const createReducer = <T extends SliceCaseReducers<ISupportProps>>(reducer: T) =
...
+    state.support.help.isHelpLoading = action.payload
   },
   setHelpMessageSuccess(state, action: IAction<boolean>) {
-    state.support.isHelpMessageSuccess = action.payload
+    state.support.help.isHelpMessageSuccess = action.payload
+  },
+  setContactUsLoading(state, action: IAction<boolean>) {
+    state.support.contactUs.isContactUsLoading = action.payload
+  },
+  setContactUsMessageSuccess(state, action: IAction<boolean>) {
```
Comment (gor017) [inline src/redux/slices/support/reducers.ts:20]: setContactUsSubmitSuccess

File: src/redux/slices/support/reducers.ts [hunk 19]
```diff
@@ -6,13 +6,19 @@ const createReducer = <T extends SliceCaseReducers<ISupportProps>>(reducer: T) =
...
 const reducers = createReducer({
   setError(state, action: IAction<null | Error | string>) {
-    state.support.error = action.payload
+    state.support.help.error = action.payload
   },
   setHelpLoading(state, action: IAction<boolean>) {
-    state.support.isHelpLoading = action.payload
+    state.support.help.isHelpLoading = action.payload
   },
   setHelpMessageSuccess(state, action: IAction<boolean>) {
```
Comment (gor017) [inline src/redux/slices/support/reducers.ts:14]: setHelpSubmitSuccess

File: src/redux/slices/support/reducers.ts [hunk 20]
```diff
@@ -6,13 +6,19 @@ const createReducer = <T extends SliceCaseReducers<ISupportProps>>(reducer: T) =
 
 const reducers = createReducer({
   setError(state, action: IAction<null | Error | string>) {
-    state.support.error = action.payload
+    state.support.help.error = action.payload
   },
   setHelpLoading(state, action: IAction<boolean>) {
```
Comment (gor017) [inline src/redux/slices/support/reducers.ts:11]: setHelpSubmittingLoading

**Review by gor017 — CHANGES_REQUESTED** [review]

File: src/manager/support/supportManager.ts [hunk 21]
```diff
@@ -1,17 +1,26 @@
...
   axiosInstance,
   sendHelpData(message: IHelpDataProps) {
-    return axiosInstance.post<IHelpDataProps, IAxiosResponse<IHelpDataProps>>(`${baseURL}`, {
+    return axiosInstance.post<IHelpDataProps, IAxiosResponse<IHelpDataProps>>(`${helpURL}`, {
       ...message,
     })
   },
+  sendContactUsData(message: IContactUsProps) {
+    return axiosInstance.post<IContactUsProps, IAxiosResponse<IHelpDataProps>>(
+      `/core/v1/${contactUsURL}`,
```
[the remark this answers is not in the document]
Comment (alexkocharyan01) [reply src/manager/support/supportManager.ts:21]: No we don't need it

File: src/manager/support/supportManager.ts [hunk 22]
```diff
@@ -1,17 +1,26 @@
...
 const axiosInstance = Axios()
 const supportManager = {
   axiosInstance,
   sendHelpData(message: IHelpDataProps) {
-    return axiosInstance.post<IHelpDataProps, IAxiosResponse<IHelpDataProps>>(`${baseURL}`, {
+    return axiosInstance.post<IHelpDataProps, IAxiosResponse<IHelpDataProps>>(`${helpURL}`, {
       ...message,
     })
   },
+  sendContactUsData(message: IContactUsProps) {
```
[the remark this answers is not in the document]
Comment (alexkocharyan01) [reply src/manager/support/supportManager.ts]: Done

File: src/manager/support/supportManager.ts [hunk 23]
```diff
@@ -1,17 +1,26 @@
...
 import { IAxiosResponse } from '@axios/axiosTypes'
 
-const baseURL = 'help'
+const helpURL = 'help'
+const contactUsURL = 'contact-us'
 
 const axiosInstance = Axios()
 const supportManager = {
   axiosInstance,
   sendHelpData(message: IHelpDataProps) {
```
[the remark this answers is not in the document]
Comment (alexkocharyan01) [reply src/manager/support/supportManager.ts]: done

File: src/redux/slices/support/middleware.ts [hunk 24]
```diff
@@ -20,10 +26,36 @@ const createHelpMessage = (message: IHelpDataProps) => async (dispatch: AppDispa
...
     dispatch(setHelpLoading(false))
-    dispatch(setHelpMessageSuccess(false))
   }
 }
 
+const resetHelpMessageSuccess = () => (dispatch: AppDispatch) => {
+  dispatch(setHelpMessageSuccess(false))
+}
+
+const createContactUsMessage = (message: IContactUsProps) => async (dispatch: AppDispatch) => {
```
[the remark this answers is not in the document]
Comment (alexkocharyan01) [reply src/redux/slices/support/middleware.ts]: done

File: src/redux/slices/support/middleware.ts [hunk 25]
```diff
@@ -1,11 +1,17 @@
...
-const { setError, setHelpLoading, setHelpMessageSuccess } = slice.actions
+const {
+  setError,
+  setHelpLoading,
+  setHelpMessageSuccess,
+  setContactUsLoading,
+  setContactUsMessageSuccess,
+} = slice.actions
 
 const createHelpMessage = (message: IHelpDataProps) => async (dispatch: AppDispatch) => {
```
[the remark this answers is not in the document]
Comment (alexkocharyan01) [reply src/redux/slices/support/middleware.ts]: done

File: src/assets/localization/en.json [hunk 26]
```diff
@@ -5,6 +5,18 @@
...
+  "contact.us.full.name": "Full Name",
+  "contact.us.full.name.required": "Full name is required",
+  "contact.us.google.button": "Open Google Maps",
+  "contact.us.mail": "E-mail",
+  "contact.us.mail.required": "Email is required",
+  "contact.us.message": "Message",
+  "contact.us.message.required": "Message Number is required",
+  "contact.us.phone": "Telephone",
+  "contact.us.phone.required": "Phone Number is required",
   "help.button": "Submit",
```
[the remark this answers is not in the document]
Comment (alexkocharyan01) [reply src/assets/localization/en.json:16]: Done

File: src/validation/support/contactUs.ts [hunk 27]
```diff
@@ -0,0 +1,8 @@
+import { object, string } from 'yup'
+
+export const contactUsValidationSchema = object({
+  fullName: string().required('Full name is required'),
```
[the remark this answers is not in the document]
Comment (alexkocharyan01) [reply src/validation/support/contactUs.ts]: done

File: src/redux/slices/support/selector.ts (hunk 11 above)
Comment (alexkocharyan01) [reply src/redux/slices/support/selector.ts]: done

File: src/redux/slices/support/selector.ts (hunk 12 above)
Comment (alexkocharyan01) [reply src/redux/slices/support/selector.ts]: done

File: src/types/reduxTypes/supportStateTypes.ts (hunk 14 above)
Comment (alexkocharyan01) [reply src/types/reduxTypes/supportStateTypes.ts]: done

File: src/redux/slices/support/middleware.ts (hunk 16 above)
Comment (alexkocharyan01) [reply src/redux/slices/support/middleware.ts]: done

File: src/redux/slices/support/middleware.ts (hunk 17 above)
Comment (alexkocharyan01) [reply src/redux/slices/support/middleware.ts]: done

File: src/redux/slices/support/reducers.ts (hunk 18 above)
Comment (alexkocharyan01) [reply src/redux/slices/support/reducers.ts]: done

File: src/validation/support/contactUs.ts [hunk 28]
```diff
@@ -0,0 +1,13 @@
+import { object, string } from 'yup'
+
+export const contactUsValidationSchema = object({
+  fullName: string()
+    .matches(/^[aA-zZ\s]+$/, 'Only alphabets are allowed for this field ')
+    .required('Full name is required'),
+  email: string()
+    .email()
+    .matches(/^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/, 'Invalid name address')
```
Comment (gor017) [inline src/validation/support/contactUs.ts:9]: Change to Invalid email address

File: src/validation/support/contactUs.ts (hunk 10 above)
Comment (gor017) [reply src/validation/support/contactUs.ts:11]: Fix this

**Review by gor017 — CHANGES_REQUESTED** [review]

**Review by gor017 — APPROVED** [review]

File: src/validation/support/contactUs.ts (hunk 28 above)
Comment (alexkocharyan01) [reply src/validation/support/contactUs.ts:9]: done

File: src/validation/support/contactUs.ts (hunk 10 above)
Comment (alexkocharyan01) [reply src/validation/support/contactUs.ts]: done

**Review by gor017 — APPROVED** [review]

**Review by ArthurO2002 — APPROVED** [review]

**Review by gor017 — APPROVED** [review]

File: src/pages/contact-us/index.tsx [hunk 29]
```diff
@@ -1,95 +1,71 @@
 import React from 'react'
+import { useTranslation } from 'react-i18next'
+import ContactUsSendMessage from '@components/ContactUsSendMessage'
```
Comment (rikarsen) [inline src/pages/contact-us/index.tsx:3]: make index.tsx file for this component

**Review by rikarsen — CHANGES_REQUESTED** [review]

File: src/pages/contact-us/index.tsx (hunk 29 above)
Comment (gor017) [reply src/pages/contact-us/index.tsx:3]: Also, change component name because it sends not only a message

Comment (alexkocharyan01) [reply src/pages/contact-us/index.tsx]: done

**Review by rikarsen — APPROVED** [review]

**Pull request merged**